### PR TITLE
[BEAM-8224] Fix bug in _fn_takes_side_inputs

### PR DESCRIPTION
**Pull request by udim**
Bug was introduced in the conversion to inspect.signature, and manifests
as assigning a wrapper function that accepts a single positional
argument when the wrapped function can accept an arbitrary number.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by aaltay** [comment]
Changes LGTM. 

Py3.7 tests are failing due to: https://issues.apache.org/jira/browse/BEAM-8229

**Comment by udim** [comment]
Run Python 3.7 PostCommit

**Comment by udim** [comment]
Run Python 3.7 PostCommit

File: sdks/python/apache_beam/typehints/typed_pipeline_test.py [hunk 1]
```diff
@@ -221,9 +221,46 @@ def repeat(s, *times):
...
+      with self.assertRaisesRegexp(
+          typehints.TypeCheckError,
+          r'requires Tuple\[int, ...\] but got Tuple\[str, ...\]'):
+        ['a', 'bb', 'c'] | beam.Map(repeat, 'z')
+
+  def test_var_positional_only_side_input_hint(self):
+    # Test that a lambda that accepts only a VAR_POSITIONAL can accept
+    # side-inputs.
+    result = (['a', 'b', 'c']
+              | beam.Map(lambda *_: 'a', 5).with_input_types(str, int))
```
Comment (robertwb) [inline sdks/python/apache_beam/typehints/typed_pipeline_test.py]: Let's check that it actually gets the values right, e.g. do `lambda *args: args`. (Similarly below.)

**Comment by markflyhigh** [comment]
https://github.com/apache/beam/pull/9570 is merge which should unblock Python2_PC failure.

**Comment by udim** [comment]
Run Python 3.7 PostCommit

**Comment by robertwb** [comment]
Run Portable_Python PreCommit

**Comment by aaltay** [comment]
New tests are failing in the precommit tests:

17:59:06 ======================================================================
17:59:06 ERROR: test_var_positional_only_side_input_hint (apache_beam.typehints.typed_pipeline_test.SideInputTest)
17:59:06 ----------------------------------------------------------------------
17:59:06 Traceback (most recent call last):
...
17:59:06 TypeError: Expected bytes, got tuple [while running 'Map(<lambda at typed_pipeline_test.py:234>)']

**Comment by robinyqiu** [comment]
Run Python PreCommit

**Comment by udim** [comment]
trivial_inference.py defaults to Any in 3.6, 3.7 since it doesn't support kwargs (added by the wrapper).
In 2.7, 3.5 these are supported, but since side-input hints are not passed along the inferred output type is str, and the coder for that fails to encode a tuple.

Even if we did pass side-input hints, there is still missing logic that would bind the arguments for the var_positional into a single tuple (like getcallargs or signature.bind).

Perhaps we should skip trivial_inference when side-inputs are present.

**Comment by aaltay** [comment]
> trivial_inference.py defaults to Any in 3.6, 3.7 since it doesn't support kwargs (added by the wrapper).
> In 2.7, 3.5 these are supported, but since side-input hints are not passed along the inferred output type is str, and the coder for that fails to encode a tuple.
> 
> Even if we did pass side-input hints, there is still missing logic that would bind the arguments for the var_positional into a single tuple (like getcallargs or signature.bind).
> 
> Perhaps we should skip trivial_inference when side-inputs are present.

Would we need to skip it for all python version or for some python versions? 

I am in favor of skipping in order to not block the release and continue with the proper fix in master.

/cc @tvalentyn in case anything needs to be documented as known issues for python 3.

**Comment by markflyhigh** [comment]
Run Portable_Python PreCommit

**Comment by udim** [comment]
Run Portable_Python PreCommit

**Comment by udim** [comment]
I've opened a couple of bug for the portable precommit flakes:
https://issues.apache.org/jira/browse/BEAM-8248
https://issues.apache.org/jira/browse/BEAM-8249

**Comment by udim** [comment]
I believe we can safely skip the portable python precommit failure

**Pull request merged**